sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1679,7 +1694,31 @@ class Analyzer(
     }
 
     def containsAggregate(condition: Expression): Boolean = {
-      condition.find(_.isInstanceOf[AggregateExpression]).isDefined
+      condition.find(e =>
+        e.isInstanceOf[AggregateExpression] ||
+          e.isInstanceOf[GroupingID] || e.isInstanceOf[Grouping]).isDefined
+    }
+
+    private def pushDownMissingAttrs(
+        missingAttrs: Seq[NamedExpression], plan: LogicalPlan): LogicalPlan = {
+      // Missing attributes can be unresolved attributes or resolved attributes which are not in
+      // the output attributes of the plan.
+      plan match {
+        case p: Project =>
+          // Recursively pushing down expressions on the child of current plan.
+          val newChild = pushDownMissingAttrs(missingAttrs, p.child)
+          Project(p.projectList ++ missingAttrs.map(_.toAttribute), newChild)
+
+        case a @ Aggregate(_, aggExprs, _) =>
+          a.copy(aggregateExpressions = aggExprs ++ missingAttrs)
+
+        // For other operators (eg. Filter), push down recursively
+        case n: UnaryNode =>
```

Comment:
Are there other operators than `Filter` can be here? If no, should we just reduce the case to `Filter`?

Reply:
I don't think so. I always saw this case with HAVING which introduces a Filter and Project... I reduce this case to Filter if you see any value in doing that (honestly I don't see)

Reply:
yes, I think so (in the HAVING case it is so at least), but I see no harm in having `UnaryNode`: I see no value in restricting this to `Filter`. But if you see some benefit I can do that.

Reply:
It sounds too ambiguous to have `UnaryNode` here because it doesn't define clearly what this operator will do. If there is an `UnaryNode` changing outputs from its child, unexpected result will get. It sounds much safer to write `Filter` here, except for that we are 100% sure that `UnaryNode` won't change its child's output.